packages/ui/src/views/entity-reference/entity-reference-component.tsx

```diff
@@ -0,0 +1,217 @@
+
+export interface EntityReferenceProps<T extends BaseEntityProps, S = string, F = string> {
+  // Data
+  entities: T[]
+  selectedEntity: T | null
+  parentFolder: S | null
+  childFolder: F | null
+
+  // Callbacks
+  onSelectEntity: (entity: T) => void
+  onScopeChange: (direction: DirectoryType) => void
+
+  // UI Configuration
+  showFilter?: boolean
+
+  // Custom renderers
+  renderEntity?: (props: EntityRendererProps<T>) => React.ReactNode
+  renderScopeSelector?: (props: ScopeSelectorProps<S>) => React.ReactNode
+  renderFolder?: (props: FolderRendererProps<F>) => React.ReactNode
+  isLoading?: boolean
+}
+
+export function EntityReference<T extends BaseEntityProps, S = string, F = string>({
+  // Data
+  entities,
+  selectedEntity,
+  parentFolder,
+  childFolder,
+
+  // Callbacks
+  onSelectEntity,
+  onScopeChange,
+
+  showFilter = true,
+
+  // Custom renderers
+  renderEntity,
+  renderScopeSelector,
+  renderFolder,
+  isLoading = false
+}: EntityReferenceProps<T, S, F>): JSX.Element {
+  const handleSelectEntity = useCallback(
+    (entity: T) => {
+      onSelectEntity?.(entity)
+    },
+    [onSelectEntity]
+  )
+
+  const handleScopeChange = useCallback(
+    (direction: DirectoryType) => {
+      onScopeChange?.(direction)
+    },
+    [onScopeChange]
+  )
+
+  const defaultEntityRenderer = ({ entity, isSelected, onSelect }: EntityRendererProps<T>) => {
+    return (
+      <StackedListItem
+        onClick={() => onSelect?.(entity)}
+        className={cn(isSelected && 'bg-background-4')}
+        thumbnail={<Icon name="file" size={16} className="text-foreground-5" />}
+        actions={
+          <Button
+            variant="default"
+            size="sm"
+            onClick={e => {
+              e.stopPropagation()
+              onSelect?.(entity)
+            }}
+          >
+            Select
+          </Button>
+        }
+      >
+        <StackedListField title={entity.name} />
+      </StackedListItem>
+    )
+  }
+
+  // Default scope selector renderer
+  const defaultScopeSelectorRenderer = ({ parentFolder, onSelect }: ScopeSelectorProps<S>) => {
+    return (
+      <StackedListItem
+        onClick={() => onSelect?.(parentFolder)}
+        thumbnail={<Icon name="circle-arrow-top" size={16} className="text-foreground-5" />}
+      >
+        <StackedListField title={<span className="capitalize">{String(parentFolder)}</span>} />
+      </StackedListItem>
+    )
+  }
+
+  // Default folder renderer
+  const defaultFolderRenderer = ({ folder, onSelect }: FolderRendererProps<F>) => {
+    return (
+      <StackedListItem
+        onClick={() => onSelect?.(folder)}
+        thumbnail={<Icon name="folder" size={16} className="text-foreground-5" />}
+      >
+        <StackedListField title={<span className="capitalize">{String(folder)}</span>} />
+      </StackedListItem>
+    )
+  }
+
+  const renderCombinedList = () => {
+    return (
+      <StackedList>
+        {/* scopes */}
+        {parentFolder && (
```

Comment:
generally better to avoid `&&` syntax
```suggestion
        { parentFolder ? (...) : null }
```